### PR TITLE
llm.generate fixes

### PR DESCRIPTION
**Pull request by HuiyingLi**
# What does this PR do ?

Apply the following fixes to llm.generate

- load from context subdir instead of ckpt dir
- disabling initializing optimizer
- add logic to find exact McoreGPTModel
- require user to provide their own trainer for llm.generate

Usage:
```
    strategy = nl.MegatronStrategy(tensor_model_parallel_size=tp, init_model_parallel=False)
    trainer = nl.Trainer(devices=tp, strategy=strategy)
    print(
        llm.generate(
            ckpt_path,
            tensor_parallel_size=2,
            prompts=["Deep learning is"],
            inference_params=CommonInferenceParams(num_tokens_to_generate=20),
        )
    )
```
Example output:
```
[InferenceRequest(request_id='0', prompt='Deep learning is', inference_parameters=CommonInferenceParams(temperature=1.0, top_k=0, top_p=0.0, return_log_probs=False, num_tokens_to_generate=20), prompt_tokens=[34564, 6975, 374, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001, 128001], arrival_time=1729579214.2338436, status=<Status.COMPLETED: 4>, encoder_prompt=None, generated_text=' the topic of hot interest nowadays. Thanks to its unique representatives including computer vision, natural language program etc', generated_tokens=tensor([  279,  8712,   315,  4106,  2802, 46877,    13, 11361,   311,  1202,
         5016, 24005,  2737,  6500, 11376,    11,  5933,  4221,  2068,  5099],
       device='cuda:0'), generated_log_probs=None, generated_length=20)]
```

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/api.py [hunk 1]
```diff
@@ -437,6 +437,7 @@ def generate(
     path: Union[Path, str],
     prompts: list[str],
     trainer: Optional[nl.Trainer] = None,
+    tensor_parallel_size: int = -1,
```
Comment (hemildesai) [inline nemo/collections/llm/api.py:440]: This default should be 1

File: nemo/collections/llm/inference/base.py [hunk 2]
```diff
@@ -62,19 +64,23 @@ def _setup_trainer_and_restore_model(path: Path, trainer: nl.Trainer, model: pl.
...
     inference_batch_times_seqlen_threshold: int = 1000,
 ) -> tuple[MCoreGPTModel, MCoreTokenizerWrappper]:
-    model: io.TrainerContext = io.load_context(path=path, subpath="model")
-    trainer = trainer or io.load_context(path=path, subpath="trainer")
+    model: io.TrainerContext = io.load_context(path=ckpt_to_context_subdir(path), subpath="model")
+    trainer = trainer or io.load_context(path=ckpt_to_context_subdir(path), subpath="trainer")
+    if tensor_parallel_size > 0:
+        trainer.strategy.tensor_model_parallel_size = tensor_parallel_size
+        trainer.devices = tensor_parallel_size
+        trainer.strategy.parallel_devices = [torch.device(f"cuda:{i}") for i in range(tensor_parallel_size)]
```
Comment (hemildesai) [inline nemo/collections/llm/inference/base.py:76]: Do we need this? I don't think we need to set this explicitly.

File: nemo/collections/llm/api.py (hunk 1 above)
Comment (hemildesai) [inline nemo/collections/llm/api.py:440]: Can we rename this to `override_tensor_parallel_size`?

File: nemo/collections/llm/inference/base.py (hunk 2 above)
Comment (HuiyingLi) [reply nemo/collections/llm/inference/base.py:76]: If we don't set explicitly, the tp will be inherit from load_context's trainer, which will be SFT training setting, which is greater than what we need when just doing llm.generate. Do we want user to be able to set it?

File: nemo/collections/llm/inference/base.py [hunk 3]
```diff
@@ -62,19 +64,23 @@ def _setup_trainer_and_restore_model(path: Path, trainer: nl.Trainer, model: pl.
...
+        trainer.strategy.launcher.num_processes = len(trainer.strategy.parallel_devices)
     _setup_trainer_and_restore_model(path=path, trainer=trainer, model=model)
 
-    # This is to get the MCore model required in GPTInferenceWrapper.
```
[the remark this answers is not in the document]
Comment (HuiyingLi) [reply nemo/collections/llm/inference/base.py:72]: I find that top level model is also McoreGPTModel and can directly be pass in and generate without problem, but I might be missing something. Do you have suggestions? I can add it back with the logic of finding the correct McoreGPTModel without subclasssing

[the remark this answers is not in the document]
Comment (hemildesai) [reply nemo/collections/llm/inference/base.py:72]: I think the class expects an Mcore object so its better to pass that to be compatible with Mcore's API.

File: nemo/collections/llm/inference/base.py (hunk 2 above)
Comment (hemildesai) [reply nemo/collections/llm/inference/base.py:76]: Oh I meant just for `trainer.strategy.parallel_devices`

File: nemo/collections/llm/inference/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HuiyingLi) [reply nemo/collections/llm/inference/base.py:72]: Sounds good! I will add the logic for finding the exact McoreGPTModel class thanks for the suggestion!

File: nemo/collections/llm/inference/base.py (hunk 2 above)
Comment (HuiyingLi) [reply nemo/collections/llm/inference/base.py:76]: If I don't change strategy.parallel_devices, both strategy.parallel_devices and launcher.num_processes will be the SFT training setting--it is init/built at load_context time. When using torchrun, the launcher will launch SFT tp setting and complain about world size mismatch. Do you think it's ok or too hacky and maybe directly modify the saved SFT context?

**Review by hemildesai — APPROVED** [review]
